lib/Agrammon/OutputFormatter/CSV.pm6

```diff
@@ -1,51 +1,94 @@
+        when Hash {
+            for sorted-kv($_) -> $output, $value {
+                next unless $model.should-print($module, $output, @print-set);
+
+                my $filters = '';
+                my $raw-value = $value;
+                my $var = $output;
+                my $unit = $model.output-unit($module, $output, $language);
                 my @data = (
-                    $module, .key, flat-value(.value) // '',
-                    $model.output-unit($module, .key, $unit-language)
+                    |($prefix if $prefix),
+                    $module,
+                    $var,
+                    $filters,
+                    flat-value($raw-value) // '',
+                    $unit
                 );
-                @data.unshift($prefix) if $prefix;
                 take @data.join(';');
+                if $include-filters {
+                    if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                        add-filters($raw-value, $module, $var, $unit, $prefix, :$all-filters);
+                    }
+                }
             }
-         }
-        else {
-            for .value.sort(*.key) {
-                my $instance-id = .key;
-                for .value.sort(*.key) {
-                    my $fq-name = .key;
+        }
+        when Array {
+            for sorted-kv($_) -> $instance-id, %instance-outputs {
+                my $filters = '';
+                for sorted-kv(%instance-outputs) -> $fq-name, %values {
                     my $q-name = $module ~ '[' ~ $instance-id ~ ']' ~ $fq-name.substr($module.chars);
-                    for .value.sort(*.key) {
+                    for sorted-kv(%values) -> $output, $value {
+                        next unless $model.should-print($module, $output, @print-set);
+
+                        my $raw-value = $value;
+                        my $unit = $model.output-unit($module, $output, $language);
                         my @data = (
-                            $q-name, .key, flat-value(.value) // '',
-                            $model.output-unit($fq-name, .key, $unit-language)
+                            |($prefix if $prefix),
+                            $q-name,
+                            $output,
+                            $filters,
+                            flat-value($raw-value) // '',
+                            $unit
                         );
-                        @data.unshift($prefix) if $prefix;
                         take @data.join(';');
+                        if $include-filters {
+                            if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                                add-filters($raw-value, $q-name, $fq-name, $unit, $prefix, :$all-filters);
+                            }
+                        }
                     }
                 }
             }
         }
-    }).join("\n") ~ "\n";
+    }).join("\n");
```

Comment:
Is everything upstream from this OK about losing the trailing `\n`?

Reply:
(For example, we don't get a regression anywhere in output writing of multiple datasets?)

Reply:
I removed it because it gave a newline too many because we output with `say`.

Reply:
Aha, makes sense.